File: 백준/Platinum/5746. Onion Layers/Onion Layers.py
import sys
input=sys.stdin.readline
def ccw(a,b,c):return(b[0]-a[0])*(c[1]-a[1])-(b[1]-a[1])*(c[0]-a[0])<0
while True:
 n=int(input())
 if not n:break
 P=sorted([(*map(int,input().split()),i)for i in range(n)])
 M=[0]*n;C=1
 while len(P)>=3:
  m=len(P);L=[];U=[]
  for j in P:
   while len(L)>1 and ccw(L[-2],L[-1],j):L.pop()
   L.append(j)
  for j in P[::-1]:
   while len(U)>1 and ccw(U[-2],U[-1],j):U.pop()
   U.append(j)
  for j in L+U:M[j[2]]=C
  P=[j for j in P if M[j[2]]!=C]
  C+=1
 print("Take this onion to the lab!"if C%2==0 else"Do not take this onion to the lab!")

File: 백준/Platinum/5746. Onion Layers/Onion Layers.cc
#include<bits/stdc++.h>
using namespace std;
struct pt{ int x,y,id; };
int ccw(pt a,pt b,pt c){ return (((b.x-a.x)*(c.y-a.y)-(b.y-a.y)*(c.x-a.x))<0); }
main(){
ios::sync_with_stdio(0),cin.tie(0),cout.tie(0);
int n;
while(cin>>n&&n){
vector<pt> P(n);
for(int i=0;i<n;i++){
cin>>P[i].x>>P[i].y;
P[i].id=i;
}
sort(P.begin(),P.end(),[&](pt a,pt b){
if(a.x==b.x) return a.y<b.y;
return a.x<b.x;
});
vector<int> mark(n,0);
int cnt=1;
while(P.size()>=3){
int m=P.size();
vector<pt> L,U,N;
N.reserve(m),L.reserve(m),U.reserve(m);
for(auto i:P){
while(L.size()>1&&ccw(L[L.size()-2],L.back(),i)) L.pop_back();
L.push_back(i);
}
for(int i=m-1;i>=0;i--){
while(U.size()>1&&ccw(U[U.size()-2],U.back(),P[i])) U.pop_back();
U.push_back(P[i]);
}
for(auto i:L) mark[i.id]=cnt;
for(auto i:U) mark[i.id]=cnt;
for(auto p:P) if(mark[p.id]!=cnt) N.push_back(p);
P.swap(N);
cnt++;
}
if(cnt&1) cout<<"Do not take this onion to the lab!\n";
else cout<<"Take this onion to the lab!\n";
}
}